src/expr/impl/src/aggregate/approx_percentile.rs

```diff
@@ -0,0 +1,139 @@
+///
+/// query R
+/// select percentile_cont(0.45) within group (order by x desc) from t;
+/// ----
+/// 2.1
+///
+/// query R
+/// select percentile_cont(0.45) within group (order by y desc) from t;
+/// ----
+/// 21
+///
+/// query R
+/// select percentile_cont(0.45) within group (order by z desc) from t;
+/// ----
+/// 210
+///
+/// query R
+/// select percentile_cont(0.45) within group (order by w desc) from t;
+/// ----
+/// 2100
+///
+/// query R
+/// select percentile_cont(NULL) within group (order by w desc) from t;
+/// ----
+/// NULL
+///
+/// statement ok
+/// drop table t;
+/// ```
+#[build_aggregate("approx_percentile(float8) -> float8")]
+fn build(agg: &AggCall) -> Result<Box<dyn AggregateFunction>> {
+    let fraction = agg.direct_args[0]
+        .literal()
+        .map(|x| (*x.as_float64()).into());
+    Ok(Box::new(ApproxPercentile { fraction }))
+}
+
+pub struct ApproxPercentile {
+    fraction: Option<f64>,
+}
+
+#[derive(Debug, Default, EstimateSize)]
+struct State(Vec<f64>);
+
+impl AggStateDyn for State {}
+
+impl ApproxPercentile {
+    fn add_datum(&self, state: &mut State, datum_ref: DatumRef<'_>) {
+        if let Some(datum) = datum_ref.to_owned_datum() {
+            state.0.push((*datum.as_float64()).into());
+        }
+    }
+}
+
+#[async_trait::async_trait]
+impl AggregateFunction for ApproxPercentile {
+    fn return_type(&self) -> DataType {
+        DataType::Float64
+    }
+
+    fn create_state(&self) -> Result<AggregateState> {
+        Ok(AggregateState::Any(Box::<State>::default()))
+    }
+
+    async fn update(&self, state: &mut AggregateState, input: &StreamChunk) -> Result<()> {
+        let state = state.downcast_mut();
+        for (_, row) in input.rows() {
+            self.add_datum(state, row.datum_at(0));
+        }
+        Ok(())
+    }
+
+    async fn update_range(
+        &self,
+        state: &mut AggregateState,
+        input: &StreamChunk,
+        range: Range<usize>,
+    ) -> Result<()> {
+        let state = state.downcast_mut();
+        for (_, row) in input.rows_in(range) {
+            self.add_datum(state, row.datum_at(0));
+        }
+        Ok(())
+    }
+
+    async fn get_result(&self, state: &AggregateState) -> Result<Datum> {
+        let state = &state.downcast_ref::<State>().0;
+        Ok(
+            if let Some(fraction) = self.fraction
+                && !state.is_empty()
+            {
+                let rn = fraction * (state.len() - 1) as f64;
+                let crn = f64::ceil(rn);
+                let frn = f64::floor(rn);
+                let result = if crn == frn {
+                    state[crn as usize]
+                } else {
+                    (crn - rn) * state[frn as usize] + (rn - frn) * state[crn as usize]
+                };
+                Some(result.into())
+            } else {
+                None
+            },
+        )
+    }
+}
```

Comment:
Can ignore the internal function implementations here. They are just placeholders.

Reply:
This sounds quite misleading to me.

Reply:
Okay I will just replace them all with todos!() they are planned to be worked on in a future PR.